pkg/agent/agent.go

```diff
@@ -29,21 +34,231 @@ import (
+
+		s.MetaPipelines[metaKey] = &NimbessPipeline{Name: fmt.Sprintf("%s_meta_pipeline", metaKey),
+			MetaKey: metaKey, Modules: make([]network.PipelineModule, 0),
+			EgressPorts: make([]*network.EgressPort, 0)}
+		if err := s.MetaPipelines[metaKey].Init("", L2DriverMode, nil); err != nil {
+			log.Error("Error while initializing MetaPipeline: %v", err)
+			return &cni.CNIReply{Result: CniIncompatible}, err
+		}
+
+	}
+
+	// Init Port Pipeline
+	s.Pipelines[portName] = &NimbessPipeline{Mu: s.Mu, Driver: s.Driver, MetaKey: metaKey,
+		Name: fmt.Sprintf("%s_pipeline", portName), Modules: make([]network.PipelineModule, 0),
+		EgressPorts: make([]*network.EgressPort, 0)}
+	if err := s.Pipelines[portName].Init(portName, L2DriverMode, s.MetaPipelines[metaKey]); err != nil {
+		log.Error("Error while initializing Port Pipeline: %v", err)
+		return &cni.CNIReply{Result: CniIncompatible}, err
+	}
+
+	// Create Port in Nimbess Pipeline
+	eModule, err := s.Pipelines[portName].AddPort(portName, port)
+	if err != nil {
+		log.Errorf("Error while adding port %v to pipeline", port)
+		return &cni.CNIReply{}, err
+	}
+
+	log.Debugf("Rendering pipeline for port: %s", portName)
+	// Render Pipeline
+	if err := s.Driver.RenderModules(s.Pipelines[portName].Modules); err != nil {
+		return &cni.CNIReply{Result: DriverFailure}, err
+	}
+
+	log.Debug("Updating EgressPorts for this pipeline")
+	// Add all current MetaPipeline Egress ports to this port pipeline
+	for _, ePort := range s.MetaPipelines[metaKey].EgressPorts {
+		if err := s.connectEgressPort(ePort, s.Pipelines[portName]); err != nil {
+			return &cni.CNIReply{}, err
+		}
+		// Add FIB entry for this port
+		if err := s.updateFIB(s.Pipelines[portName], *ePort); err != nil {
+			return &cni.CNIReply{}, err
+		}
+	}
+
+	log.Debug("Updating EgressPorts for other pipelines")
+	// Update all other Pipelines with new Egress Port, including MetaPipeline
+	if err := s.updatePipelinesEgress(eModule, portName, s.Pipelines[portName].MetaKey); err != nil {
+		return &cni.CNIReply{}, err
+	}
+
+	cniIfIP := &cni.CNIReply_Interface_IP{
+		Version: cni.CNIReply_Interface_IP_IPV4,
+		Address: ipAddr.String(),
+	}
+
+	cniIf := &cni.CNIReply_Interface{
+		Name:        portName,
+		Mac:         port.MacAddr,
+		IpAddresses: []*cni.CNIReply_Interface_IP{cniIfIP},
+	}
+
+	return &cni.CNIReply{
+		Result:     CniOk,
+		Error:      "",
+		Interfaces: []*cni.CNIReply_Interface{cniIf},
+	}, nil
+}
+
+// updateL2FIB adds a new FIB entry to a Pipeline's forwarder
+func (s *NimbessAgent) updateL2FIB(l2Forwarder *network.Switch, macAddr string, gate network.Gate) error {
+	if err := s.Driver.AddEntryL2FIB(l2Forwarder, macAddr, gate); err != nil {
+		return err
+	}
+	l2Forwarder.L2FIB[macAddr] = gate
+	log.Infof("Added L2FIB entry: %s:%d to switch: %s", macAddr, gate, l2Forwarder.Name)
+	return nil
+}
+
+// updateFIB adds a new FIB entry to a Pipeline's forwarder for a new EgressPort
```

Comment:
guess updateL2FIB and updateFIB are intuitive to know one is the pipeline forwarder and the other is with egressport?

Reply:
updateFIB is supposed to be just the generic/abstract function that handles figuring out what type of FIB exists, which module it is, and then passing it to the more specific updateL2FIB or updateL3FIB. I just didn't implement L3 FIB yet because we haven't gotten that far yet.